var/spack/repos/builtin/packages/hpx/package.py

```diff
@@ -71,20 +72,72 @@ class Hpx(CMakePackage, CudaPackage):
     variant('async_mpi', default=False, description='Enable MPI Futures.')
     variant('async_cuda', default=False, description='Enable CUDA Futures.')
 
-    depends_on('hwloc')
+    # Build dependencies
     depends_on('python', type=('build', 'test', 'run'))
+    depends_on('ninja', type='build')
     depends_on('pkgconfig', type='build')
     depends_on('git', type='build')
+    depends_on('cmake', type='build')
+
+    # Other dependecies
+    depends_on('hwloc')
+    depends_on('boost')
+    for cxxstd in cxxstds:
+        depends_on(
+            "boost cxxstd={0}".format(map_cxxstd(cxxstd)), when="cxxstd={0}".format(cxxstd)
+        )
+    depends_on('asio', when='@1.7:')
+    for cxxstd in cxxstds:
+        depends_on(
+            "asio cxxstd={0}".format(map_cxxstd(cxxstd)),
+            when="cxxstd={0} ^asio".format(cxxstd),
+        )
+
+    depends_on('gperftools', when='malloc=tcmalloc')
+    depends_on('jemalloc', when='malloc=jemalloc')
+    depends_on('tbb', when='malloc=tbbmalloc')
+
+    depends_on('mpi', when='networking=mpi')
+    depends_on('mpi', when='+async_mpi')
+
+    depends_on('cuda', when='+async_cuda')
+
+    depends_on('otf2', when='instrumentation=apex')
+    depends_on('gperftools', when='instrumentation=google_perftools')
+    depends_on('papi', when='instrumentation=papi')
+    depends_on('valgrind', when='instrumentation=valgrind')
+
+    # Restrictions for stable/master
+    conflicts('cxxstd=14', when='@master')
+    conflicts('cxxstd=14', when='@stable')
+    depends_on('cuda@11:', when='+cuda')
+
+    # Restrictions for 1.7.X
+    depends_on('cmake@3.18.0:', when='@1.7:', type='build')
+    depends_on('boost@1.71.0:', when='@1.7:')
+    depends_on('asio@1.12.0:', when='@1.7:')
+    conflicts('%gcc@:6', when='@1.7:%gcc')
+    conflicts('%clang@:6', when='@1.7:%clang')
```

Comment:
```suggestion
    with when('@1.7:'):
        depends_on('cmake@3.18.0:', type='build')
        depends_on('boost@1.71.0:')
        depends_on('asio@1.12.0:')
        conflicts('%gcc@:6')
        conflicts('%clang@:6')
```
this is a new feature in develop (didn't test it ^^)

Reply:
Fancy. I'll try it out!

Reply:
Worked nicely, I've updated the package to use `with when` where appropriate.